### ISSUE TITLE
Koreader getting slower and slower as I add highlights

### ISSUE THREAD

**Issue by abdoulayeYATERA** [koreader/koreader#5766]
* KOReader version: v2019.12-13-g59deaf6_2020-01-02
* Device: Kobo format

#### Issue
 While reading big books, KoReader get slower and slower as i add highlights until it's barely usable anymore.
This is the folder containing the book and all my highlights and metada so far.
[Mises_human_action.zip](https://github.com/koreader/koreader/files/4058676/Mises_human_action.zip)
The book : Human action by Mises
No copyright issue, the book is free and provided by https:///mises.org

As you'll see the metadata file is quiet big .... 5000 lines, 500kb.
I have to note that if I open an other book everything is fine.
A help will be appreciated !
Thank's !

#### Steps to reproduce
- read a big book
- add many highlights

**Comment by poire-z** [comment]
Related/duplicate: #3261 #3503 (but let's keep this issue open).
#3503 is actually the exact same/duplicate of this issue, thas has been opened 2 years ago by ... *you* :) and that I had hope it was fixed by #3508 - that you confirmed :).

Do you think it has gotten worse at some point in time since?
Or it's just that this is a bigger book with very much more highlights (250) than what you usuall read/do?

Anyway, thank you for the zip with the epub AND the sdr. Never seen that many highlights :) it's good stuff to see/test what we can do.

**Comment by abdoulayeYATERA** [comment]
"the exact same/duplicate of this issue, thas has been opened 2 years ago by ... you :) " X)
Hi Poire-z ! Yeah, i thought the symptoms were the same but not the cause.
That's why I openned a new issue.

**Comment by abdoulayeYATERA** [comment]
This book contains clearly much more highlights that usually....

**Comment by abdoulayeYATERA** [comment]
And I can add this, the navigation between pages is slow but i can cope with it, however, the text selection is really tough !

**Comment by poire-z** [comment]
OK, so it's the 500 calls to `getPosFromXPointer(xp)` (to get the pages the start and the end of each of your 250 highlights xpointers/xpath are, to see if they are part of the current page and if we need to get more details about them to draw the highlights) that are happening on each page turn, that are causing your slow down.
#3508 was about doing that, using `getPosFromXPointer(xp)` to check for the page, before doing the more expensive "getting more details" work - so, it's good it is that way.
Nothing much we can do to speed up these calls, 500 x 1 or 2ms, and you get your slowdown.
But...:

One year ago, I experimented with caching in frontend the results of many crengine calls - to avoid re-doing some work when we know we would get the same results.
I wasn't impressed with the results: it worked, but the gain wasn't much.
But that means crengine is pretty fast at doing what it does :) so, good, no real need for that caching, and I dropped the idea.
(I remember I mentionned that at the time in some issue, but I can't find it - may be when @NiLuJe was working on improving blitbuffer blitting/copying - I think I was noticing that somehow, crengine drawing on the screen BB was nearly as fast as just copying a BB to the screen BB - so, there was no use to cache crengine drawing when opening/closing menus and redrawing the same thing...)

So, I just resurrected the thing, adapted it for the new stuff that has been added over the last year, and, on your book, the result is quite impressive!
Here a single file [credocument.lua.txt](https://github.com/koreader/koreader/files/4061717/credocument.lua.txt) (remove the .txt), to replace `koreader/frontend/document/credocument.lua` with (should work with your v2019.12-13 as well as more recent builds).
See how it does and helps, and tell us if you notice strange things/bugs.

In that file, there is a single function that is called at end of :init():
```lua
@@ -133,6 +133,8 @@ function CreDocument:init()
+    -- Init current page cache
+    self:viewCacheEnable(true, false)
 end
```
and this big function at bottom of file: it wraps some of the original methods into some generic wrapper - that caches the results of the original methods, and deal with dropping cache when needed.
It manages some statistics, so here are a few of them when playing with your book:

Just reading a few pages:
```
CRE view cache content:
     all: 602 entries
  global: 592 entries
    tags: 10 entries
          'page1207': 9 entries
          'page1206': 9 entries
          'page1205': 9 entries
          'page1208': 9 entries
          'page1204': 9 entries
          'page1203': 9 entries
          'page1202': 9 entries
          'page1201': 9 entries
          'page1200': 9 entries
          'page1199': 9 entries
CRE view cache hits statistics:
    getPageCount: 74/76 hits (97%) 0.003s (-0.002s) saved, 0.000s used (54% saved)
    getPageFromXPointer: 1130/1169 hits (96%) 0.759s (-0.019s) saved, 0.026s used (94% saved)
#>  getPosFromXPointer: 81950/82500 hits (99%) 139.560s (-1.312s) saved, 0.937s used (98% saved)
    getXPointer: 113/144 hits (78%) 0.096s (-0.004s) saved, 0.026s used (76% saved)
    isXPointerInCurrentPage: 336/584 hits (57%) 0.179s (-0.004s) saved, 0.132s used (56% saved)
  No hit for:
    drawCurrentView: 72 misses 4.058s
    getCacheFilePath: 1 misses 0.000s
  By call times (hits vs misses):
    getPageCount: (74) 0.028 ms vs 0.035 ms (2)
    getPageFromXPointer: (1130) 0.017 ms vs 0.672 ms (39)
#>  getPosFromXPointer: (81950) 0.016 ms vs 1.703 ms (550)
    getXPointer: (113) 0.035 ms vs 0.848 ms (31)
    isXPointerInCurrentPage: (336) 0.013 ms vs 0.534 ms (248)
  cpu time used: 6.521s, saved: 139.256s (95% saved)
```
Reading a few pages, highlighting, changing font sizes, looking at footnotes:
```
CRE view cache content:
     all: 570 entries
  global: 566 entries
    tags: 4 entries
          'page1178': 9 entries
          'page1179': 9 entries
          'page1177': 9 entries
          'page1180': 9 entries
CRE view cache hits statistics:
    drawCurrentView: 13/97 hits (13%) 0.648s (-0.168s) saved, 4.189s used (12% saved)
    getHeaderHeight: 7/8 hits (87%) 0.000s saved, 0.000s used (69% saved)
    getLinkFromPosition: 5/12 hits (41%) 0.003s saved, 0.004s used (37% saved)
    getPageCount: 59/62 hits (95%) 0.004s (-0.005s) saved, 0.000s used (46% saved)
    getPageFromXPointer: 943/1263 hits (74%) 1.389s (-0.032s) saved, 0.471s used (73% saved)
    getPageLinks: 6/10 hits (60%) 0.016s saved, 0.010s used (58% saved)
    getPageMargins: 20/21 hits (95%) 0.002s saved, 0.000s used (79% saved)
#>  getPosFromXPointer: 79332/80448 hits (98%) 129.557s (-1.533s) saved, 1.823s used (97% saved)
    getScreenBoxesFromPositions: 83/102 hits (81%) 3.572s (-0.004s) saved, 0.818s used (81% saved)
    getScreenPositionFromXPointer: 7/15 hits (46%) 0.004s saved, 0.004s used (44% saved)
    getXPointer: 77/108 hits (71%) 0.058s (-0.004s) saved, 0.023s used (68% saved)
    isXPointerInCurrentPage: 224/479 hits (46%) 0.202s (-0.007s) saved, 0.230s used (45% saved)
  No hit for:
    getCacheFilePath: 1 misses 0.000s
  By call times (hits vs misses):
    drawCurrentView: (13) 12.889 ms vs 49.869 ms (84)
    getHeaderHeight: (7) 0.019 ms vs 0.062 ms (1)
    getLinkFromPosition: (5) 0.127 ms vs 0.543 ms (7)
    getPageCount: (59) 0.077 ms vs 0.069 ms (3)
    getPageFromXPointer: (943) 0.034 ms vs 1.473 ms (320)
    getPageLinks: (6) 0.141 ms vs 2.611 ms (4)
    getPageMargins: (20) 0.016 ms vs 0.082 ms (1)
#>  getPosFromXPointer: (79332) 0.019 ms vs 1.633 ms (1116)
    getScreenBoxesFromPositions: (83) 0.051 ms vs 43.034 ms (19)
    getScreenPositionFromXPointer: (7) 0.048 ms vs 0.540 ms (8)
    getXPointer: (77) 0.049 ms vs 0.755 ms (31)
    isXPointerInCurrentPage: (224) 0.032 ms vs 0.901 ms (255)
  cpu time used: 9.327s, saved: 133.700s (93% saved)
```
So, as we see, the main gain is `getPosFromXPointer()` being heavily avoided, and saving 120s!


For my current book with no highlights, the result is less impressive:
```
CRE view cache content:
     all: 14 entries
  global: 4 entries
    tags: 10 entries
          'page130': 3 entries
          'page129': 1 entries
          'page128': 1 entries
          'page118': 1 entries
          'page125': 1 entries
          'page126': 1 entries
          'page136': 1 entries
          'page135': 1 entries
          'page134': 1 entries
          'page133': 1 entries
CRE view cache hits statistics:
    drawCurrentView: 8/72 hits (11%) 0.281s (-0.034s) saved, 2.249s used (10% saved)
    getPageCount: 50/52 hits (96%) 0.002s (-0.009s) saved, 0.000s used (17% saved)
    getXPointer: 81/96 hits (84%) 0.150s (-0.012s) saved, 0.028s used (79% saved)
  No hit for:
    getCacheFilePath: 1 misses 0.000s
    getLinkFromPosition: 2 misses 0.001s
    getPageFromXPointer: 1 misses 0.002s
  By call times (hits vs misses):
    drawCurrentView: (8) 4.188 ms vs 35.135 ms (64)
    getPageCount: (50) 0.178 ms vs 0.037 ms (2)
    getXPointer: (81) 0.146 ms vs 1.849 ms (15)
  cpu time used: 2.334s, saved: 0.378s (13% saved)
```

I'm not really font of having such a cache, as it adds another layer of doubt when debugging, and it may hide some crengine issues that I would otherwise (happily:) investigate.
But the results are quite impressive: 95% cpu time not wasted with that book, which means more battery life.
So I dunno, may be we could have that as an option, and/or auto-enable it under some condition (like, when there are more than 20 highlights, and others). Or always, and have an option for developpers to disable it.
What do you think?

**Comment by abdoulayeYATERA** [comment]
Wow ! Thank for your fast answer Poire-z (Tu es francophone je suppose, vu le 'Poire' et non 'Peer' ^_^").
The caching result is terrific, just great !!!
I'll test it a few days and make a feedback here.

> One year ago, I experimented with caching in frontend the results of many crengine calls - to avoid re-doing some work when we know we would get the same results.
I wasn't impressed with the results: it worked, but the gain wasn't much.
But that means crengine is pretty fast at doing what it does :) so, good, no real need for that caching, and I dropped the idea.

Indeed, i guess the result gain is non-linear/exponential as you have more highlights.
> cpu time used: 9.327s, saved: 133.700s (93% saved)

I love you my friend X)

> I'm not really font of having such a cache, as it adds another layer of doubt when debugging, and it may hide some crengine issues that I would otherwise (happily:) investigate.
But the results are quite impressive: 95% cpu time not wasted with that book, which means more battery life.
So I dunno, may be we could have that as an option, and/or auto-enable it under some condition (like, when there are more than 20 highlights, and others). Or always, and have an option for developpers to disable it.
What do you think?

Well, i'm not the best to answer technically as i have not deep hindsight of the code base. 
However, functionally, it would be great to have it by default/auto-enable, and maybe keep an option to disable it as you proposed.
Thank's again, the result it awesome, i'll make an other feedback in a couple of days.

**Comment by NiLuJe** [comment]
@poire-z: I couldn't point you to the discussion, but I do remember those comments ;).

I can't comment on the actual implementation since I'm not at my main box right now, but I think it's be a shame to drop that ;).

I do get your concerns about another layer indirection possibly hiding/causing issues, so, yeah, it probably ought to be disabled by default, and then possibly add a hint when a document hits a given highlight/note thresholds to point users to the setting.

**Comment by Monirzadeh** [comment]
> And I can add this, the navigation between pages is slow but i can cope with it, however, the text selection is really tough !

i think text selection is a little slower after [this](https://github.com/koreader/koreader/pull/5763) fix in update 2019.1.38 ([related issue](https://github.com/koreader/koreader/issues/5735))
i can't measure that so i am not completely sure about that

**Comment by poire-z** [comment]
> Tu es francophone je suppose

Yes, tu supposes bien, I'm french :)

>  it'd be a shame to drop that, it looks fairly cool
>  it probably ought to be disabled by default, and then possibly add a hint when a document hits a given highlight/note threshold to point users to the setting (and/or auto-enable it).

OK, I'll make a PR, with it disabled by default. Us and those interested can enable it to see for bugs/optimisations for a few weeks/months, and if it's all fine, we may as well enable it for all, not just books with many highlights (it might save some battery life even on small zero-highlights books). We'll ask users to disable it if we have doubts, and us devs might have to toggle it when investigating if needed.

> i think text selection is a little slower after [this](https://github.com/koreader/koreader/pull/5763) fix in update 2019.1.38 ([related issue](https://github.com/koreader/koreader/issues/5735))
i can't measure that so i am not completely sure about that

@Monirzadeh You're right, it surely is slower, but by how much, I dunno.
The previous method was cheap and dirty (which might be the right thing to be for live text selection...)
The new one does mostly for each selected line what the previous one did only once for the whole selected block, so it has indeed much more work to do. But that's needed for showing what's being selected correctly in RTL or more complex text/floats/tables layout. It has some price :)
I'm not a heavy text highlighter, so I don't have enough experience to get a real feeling if text-selection slowness is noticable. But with some quick tests on my device, it didn't feel (to me) slower, I did not notice a slowdown. And if you are _not completely sure_ , I guess it's not supernoticable :)
We can go back to the old method easily if needed - if you RTL reader and other LTR readers really notice that slowdown and feel it is annoying (or we can add an option to toggle between the two, but only if there's enough feedback requesting it).
(But if @abdoulayeYATERA is still using v2019.12-13 as mentionned in first post, he is not yet using that text selection improvement - and playing with the patched credocument.lua with his book full of highlights, it's as fast (or slow) as with a book with no highlight - so a new highlight drawing speed does not depend on the number of existing highlights, which is good :)

**Comment by poire-z** [comment]
> Us and those interested can enable it to see for bugs/optimisations for a few weeks/months, and if it's all fine, we may as well enable it for all,

OK, so it will be in tomorrow's nightly, and can be enabled with:
`File browser > Tools > Developer options > { ] Enable CRE call cache`
Please enable it and just use your device normally, until somethings looks strange :)

Adding this issue to a future milestone, so we don't forget to enable it if no feedback :)
Should be just a matter of replacing some `:isTrue()` with `:nilOrTrue()`.

**Comment by abdoulayeYATERA** [comment]
Hi ! After days with cre call cache enabled everything is working fine for me.

**Comment by poire-z** [comment]
Thanks. So, i'm going to enable it by default, so nightly builds users can join the validation force.
